tutorial/source/bayesian_regression_ii.ipynb

```diff
@@ -0,0 +1,728 @@
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Model + Guide\n",
+    "\n",
+    "We will write out the model again, similar to that in [Part I](bayesian_regression.ipynb), but explicitly without the use of `nn.Module`.  We will write out each term in the regression, using the same priors. `bA` and `bR` are regression coefficients corresponding to  `is_cont_africa` and `ruggedness`, `a` is the intercept, and `bAR` is the correlating factor between the two features.\n",
+    "\n",
+    "Writing down a guide will proceed in close analogy to the construction of our model, with the key difference that the guide parameters need to be trainable. To do this we register the guide parameters in the ParamStore using `pyro.param()`. Note the positive constraints on scale parameters."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 3,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def model(is_cont_africa, ruggedness, log_gdp):\n",
+    "    a = pyro.sample(\"a\", dist.Normal(8., 1000.))\n",
+    "    b_a = pyro.sample(\"bA\", dist.Normal(0., 1.))\n",
+    "    b_r = pyro.sample(\"bR\", dist.Normal(0., 1.))\n",
+    "    b_ar = pyro.sample(\"bAR\", dist.Normal(0., 1.))\n",
+    "    sigma = pyro.sample(\"sigma\", dist.Uniform(0., 10.))\n",
+    "    mean = a + b_a * is_cont_africa + b_r * ruggedness + b_ar * is_cont_africa * ruggedness\n",
+    "    with pyro.iarange(\"data\", len(ruggedness)):\n",
+    "        pyro.sample(\"obs\", dist.Normal(mean, sigma), obs=log_gdp)\n",
+    "        \n",
+    "def guide(is_cont_africa, ruggedness, log_gdp):\n",
+    "    a_loc = pyro.param('a_loc', torch.tensor(0.))\n",
+    "    a_scale = pyro.param('a_scale', torch.tensor(1.),\n",
+    "                         constraint=constraints.positive)\n",
+    "    sigma_loc = pyro.param('sigma_loc', torch.tensor(1.),\n",
+    "                             constraint=constraints.positive)\n",
+    "    weights_loc = pyro.param('weights_loc', torch.randn(3))\n",
+    "    weights_scale = pyro.param('weights_scale', torch.ones(3),\n",
+    "                               constraint=constraints.positive)\n",
+    "    a = pyro.sample(\"a\", dist.Normal(a_loc, a_scale))\n",
+    "    b_a = pyro.sample(\"bA\", dist.Normal(weights_loc[0], weights_scale[0]))\n",
+    "    b_r = pyro.sample(\"bR\", dist.Normal(weights_loc[1], weights_scale[1]))\n",
+    "    b_ar = pyro.sample(\"bAR\", dist.Normal(weights_loc[2], weights_scale[2]))\n",
+    "    sigma = pyro.sample(\"sigma\", dist.Normal(sigma_loc, torch.tensor(0.05)))\n",
+    "    mean = a + b_a * is_cont_africa + b_r * ruggedness + b_ar * is_cont_africa * ruggedness"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 4,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Utility function\n",
+    "def summary(traces, sites):\n",
+    "    marginal = EmpiricalMarginal(traces, sites)._get_samples_and_weights()[0].detach().cpu().numpy()\n",
```

Comment:
This can be simplified to `EmpiricalMarginal(..).enumerate_support().cpu().numpy`. 

@eb8680 - what do you think about the marginals for both SVI and MCMC to return what we currently have as `_Marginal` within MCMC (without the diagnostics which MCMC can override). While `EmpiricalMarginal` works here, it wouldn't work when latents have different shapes. Moreover, we now have this ugly behavior where `.marginal()` on SVI returns `EmpiricalMarginal` and `.marginal()` for MCMC returns an internal `_Marginal` class.

Reply:
I'm fine with updating `SVI` to match `MCMC`.  We can always revisit these APIs further after the release.

Reply:
+1, in part one i have to manually run the empirical distribution to avoid shape errors